src/storage/src/hummock/sstable/multi_builder.rs

```diff
@@ -115,202 +230,275 @@ impl<F: TableBuilderFactory> CapacitySplitTableBuilder<F> {
+pub struct LocalTableBuilderFactory<B>
+where
+    B: SstableWriterBuilder,
+{
+    next_id: AtomicU64,
+    options: SstableBuilderOptions,
+    limiter: MemoryLimiter,
+    writer_builder: B,
+}
+
+impl<B> LocalTableBuilderFactory<B>
+where
+    B: SstableWriterBuilder,
+{
+    pub fn new(next_id: u64, writer_builder: B, options: SstableBuilderOptions) -> Self {
+        Self {
+            limiter: MemoryLimiter::new(1000000),
+            next_id: AtomicU64::new(next_id),
+            options,
+            writer_builder,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<B> TableBuilderFactory<B> for LocalTableBuilderFactory<B>
+where
+    B: SstableWriterBuilder,
+{
+    async fn open_builder(&self) -> HummockResult<(MemoryTracker, SstableBuilder<B::Writer>)> {
+        let id = self.next_id.fetch_add(1, SeqCst);
+        let builder = SstableBuilder::new_for_test(
+            id,
+            self.writer_builder.build(id).await?,
+            self.options.clone(),
+        );
+        let tracker = self.limiter.require_memory(1).await.unwrap();
+        Ok((tracker, builder))
     }
 }
 
 #[cfg(test)]
 mod tests {
-    use std::sync::atomic::AtomicU64;
-    use std::sync::atomic::Ordering::SeqCst;
-
     use super::*;
     use crate::hummock::iterator::test_utils::mock_sstable_store;
     use crate::hummock::sstable::utils::CompressionAlgorithm;
     use crate::hummock::test_utils::default_builder_opt_for_test;
-    use crate::hummock::{MemoryLimiter, SstableBuilderOptions, DEFAULT_RESTART_INTERVAL};
-
-    pub struct LocalTableBuilderFactory {
-        next_id: AtomicU64,
-        options: SstableBuilderOptions,
-        limiter: MemoryLimiter,
+    use crate::hummock::{SstableBuilderOptions, DEFAULT_RESTART_INTERVAL};
+
+    fn get_builder<B, S>(
+        opt: &SstableBuilderOptions,
+        writer_builder: B,
+        builder_sealer: S,
+    ) -> CapacitySplitTableBuilder<LocalTableBuilderFactory<B>, B, S>
+    where
+        B: SstableWriterBuilder,
+        S: SstableBuilderSealer<B::Writer>,
+    {
+        let builder_factory = LocalTableBuilderFactory::new(1001, writer_builder, opt.clone());
+        CapacitySplitTableBuilder::new(builder_factory, builder_sealer)
     }
 
-    impl LocalTableBuilderFactory {
-        pub fn new(next_id: u64, options: SstableBuilderOptions) -> Self {
-            Self {
-                limiter: MemoryLimiter::new(1000000),
-                next_id: AtomicU64::new(next_id),
-                options,
-            }
-        }
-    }
-
-    #[async_trait::async_trait]
-    impl TableBuilderFactory for LocalTableBuilderFactory {
-        async fn open_builder(&self) -> HummockResult<(MemoryTracker, SstableBuilder)> {
-            let id = self.next_id.fetch_add(1, SeqCst);
-            let builder = SstableBuilder::new_for_test(id, self.options.clone());
-            let tracker = self.limiter.require_memory(1).await.unwrap();
-            Ok((tracker, builder))
+    macro_rules! run_test {
```

Comment:
Is there any reason to use a macro instead of a generic method?

Reply:
Because `test_inner` may have more arguments than just the builder.

Reply:
Seems that the input of all `test_inner` is either `(CapacitySplitTableBuilder)` or `(CapacitySplitTableBuilder, table_capacity: usize)`. If so, we may unify all `test_inner` to have `(CapacitySplitTableBuilder, table_capacity: usize)` as input, and ignore the `table_capacity` accordingly.

Reply:
If we have to pass more parameters into `test_inner` in the future, we have to modify the signature of `run_test` and ignore even more parameters, which seems a bit unclean.

Reply:
At least for now we don't have need to have more parameters. 

Even if in the future we have the need to add more parameters, we can unify the input of `test_inner` into `test_inner<E>(CapacitySplitTableBuilder, extra_para: E)` and pass the extra parameters to `extra_para`. If we have multiple parameters, we can pass a tuple into it. Using macro makes the code much less readable.